lizrd/text/packers.py

```diff
@@ -0,0 +1,206 @@
+
+class BERTPacker(
+    AbstractPacker,
+):
+    def __init__(
+        self,
+        sequence_length: int,
+        dataset: AbstractDataset,
+        tokenizer_maker: Callable[[], AbstractTokenizer],
+        mask_replace_config: MaskingReplacementConfig = MaskingReplacementConfig(),
+        seed: Optional[int] = None,
+    ):
+        super().__init__(
+            sequence_length,
+            dataset,
+            tokenizer_maker,
+            seed=seed,
+        )
+        self.mask_replace_config = mask_replace_config
+
+    def get_sample(self) -> LLMExample:
+        """
+        Sample examples from the dataset until we reach the desired sequence length.
+        """
+        target_ids: List[int] = []
+        input_ids: List[int] = []
+        calculate_loss: List[int] = []
+        document_lengths: List[int] = []
+
+        sep_id = self.tokenizer.sequence_separator_id
+        assert sep_id is not None
+
+        while True:
+            document = self.dataset.get_document()
+            tokens = self.tokenizer.text_to_ids(document)
+            masked_input, is_mask = self._mask_text(tokens)
+
+            target_ids.extend(tokens + [sep_id])
+            input_ids.extend(masked_input + [sep_id])
+            calculate_loss.extend(is_mask + [0])
+
+            document_lengths.append(len(tokens) + 1)
+            if (sum(document_lengths) - max(document_lengths)) > self.sequence_length:
+                break
+
+        sample_start = self.py_rng.randint(0, len(target_ids) - 1)
+        sample_end = sample_start + self.sequence_length
+
+        target_ids = list(take_circular(target_ids, sample_start, sample_end))
+        input_ids = list(take_circular(input_ids, sample_start, sample_end))
+        calculate_loss = list(take_circular(calculate_loss, sample_start, sample_end))
+
+        return LLMExample(input_ids, target_ids, calculate_loss)
+
+    def _mask_text(self, tokens: List[int]) -> Tuple[List[int], List[int]]:
+        mask_id = self.tokenizer.mask_id
+        assert mask_id is not None
+
+        is_mask = self.np_rng.binomial(
+            1, self.mask_replace_config.mask_percentage, len(tokens)
+        )
+        how_to_mask = self.np_rng.multinomial(
+            1,
+            [
+                self.mask_replace_config.replace_with_mask,
+                self.mask_replace_config.replace_with_random,
+                self.mask_replace_config.replace_with_original,
+            ],
+            size=len(tokens),
+        ).nonzero()[1]
+        replacements = (
+            (how_to_mask == 0) * mask_id
+            + (how_to_mask == 1) * self._get_valid_random_tokens(len(tokens))
+            + (how_to_mask == 2) * tokens
+        )
+
+        input_tokens = np.where(is_mask.astype(bool), replacements, tokens)
+
+        return input_tokens.tolist(), is_mask.tolist()
+
+    def _get_valid_random_tokens(self, tokens_count):
+        assert isinstance(self.tokenizer, BertTokenizer)
+        special_tokens = 999
```

Comment:
To wygląda na magiczną liczbę, można mieć to zdefiniwane jako macro nawet w pliku. Nie jest oczywiste, czy chodzi o liczbę specjalnych tokenów, czy że taki specjalny token ma takie id

Reply:
uppercase?

Reply:
Też tego nie rozumiem. To po prostu sprawia, że nie będziemy używać jako maski pierwszych 'special_tokens'. Jeśli już przyjąć jakiś standard, to dałbym coś w stylu 10 - raczej nie będzie więcej niż 10. Można też zostawić tak z komentarzem, że to do zmiany; to jest raczej minor.

Reply:
To jest stara linijka, skopiowana. W tym tokenizerze pierwsze 999 id jest zarezerwowane na tokeny specjalne, dlatego je trzeba skipnąć

Reply:
<img width="275" alt="image" src="https://github.com/llm-random/llm-random/assets/70006947/28237f34-981f-4047-9cfd-e8e446e08aa6">